File: Sweet.java
import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.tree.*;
import java.io.FileInputStream;
import java.io.InputStream;

public class Sweet {
    public static void main(String[] args) throws Exception {
        String inputFile = null;
        if (args.length > 0) {
            inputFile = args[0];
        }

        InputStream is = System.in;
        if (inputFile != null) {
            is = new FileInputStream(inputFile);
        }
        ANTLRInputStream input = new ANTLRInputStream(is);
        SweetLexer lexer = new SweetLexer(input);
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        SweetParser parser = new SweetParser(tokens);
        ParseTree tree = parser.program();
        System.out.println(tree.toStringTree(parser));

        FunctionTable functionTable = new FunctionTable();
        InterpreterVisitor interpreter = new InterpreterVisitor(functionTable);
        interpreter.visit(tree);

        Function main = functionTable.get("main");
        main.call();
/*
        ParseTreeWalker walker = new ParseTreeWalker(); // create standard walker 
        InterpreterListener interpreter = new InterpreterListener(parser);
        walker.walk(interpreter, tree); // initiate walk of tree with listener
*/
    }
}

File: src/main/scala/Sweet.scala
import org.antlr.v4.runtime._
import org.antlr.v4.runtime.tree._
import java.io.{ FileInputStream, InputStream, PrintStream }

object Sweet {
  private var output = System.out

  def main(args: Array[String]) {
    val is = if (args.length > 0) new FileInputStream(args(0)) else System.in
    run(is, System.out)
  }

  def run(in: InputStream, out: PrintStream) {
    this.output = out
    val input = new ANTLRInputStream(in)
    val lexer = new SweetLexer(input)
    val tokens = new CommonTokenStream(lexer)
    val parser = new SweetParser(tokens)
    val tree = parser.program
    System.out.println(tree.toStringTree(parser))

    val scope = new Scope(null)
    scope.define(new Value("print", new PrintFunction))
    scope.define(new Value("Array", ArrayObject.constructor))
    val interpreter = new InterpreterVisitor(scope)
    interpreter.visit(tree)

    val main = scope.resolve("main").asInstanceOf[Function]
    main.call(List.empty[SweetObject], List.empty[Value])
  }

  def STDOUT() = { this.output }
}
